docs/framework/20categories/30implementation/upgrade.md

```diff
@@ -1,13 +1,12 @@
 ### Upgradable
-TODO: This needs some thought.
+Like any software, smart contracts software can have bugs. This is true of both code that appears to be extremely simple and more complex code. Additionally, applications may require new features. As such, many projects have the ability to upgrade their smart contract code.
 
-Upgrading can lead to rug-pulling, possibly by attackers who gain access to the address with upgrade privileges.
-However, code often has bugs. Additionally, new features can be added. Hence, the ability to upgrade is important.
-Are different upgrade mechanisms more and less risky?
+Having the ability to upgrade a contract is inherently risky as it can lead to rug-pulling attacks, where the operators of a project change the contract logic and steal customer funds (for instance the [Hunter Defi](https://twitter.com/CertiKAlert/status/1523153063875575809) rug-pull). However, as a proportion of both the number of cypto rug pull attacks and as a proportion of the projects that use upgradable contracts, using contract upgrade as a method of stealing funds is very rare. We have no hard statistics on this; this is just an observation.
 
-Transparent proxy
+The three common methodologies for upgrading smart contracts are:
 
-Poor implementation where storage slots are mistakenly re-used has led to vulnerabilities.
-Deploy new contract and point front end at new contract.
-Non-transparent proxy with separate data contract.
+* [Data Holder Upgrade Pattern](https://www.youtube.com/watch?v=VhzafmGGmzo): Have a data holder contract and a separate business logic contract. The business logic contracts are upgraded, and connect to the existing data holder contracts. Issues with this approach are that new versions of the business logic contract need to be able to utilise old data formats stored in the data holder contract.
+* [Transparent Upgrade Proxy Pattern](https://docs.openzeppelin.com/upgrades-plugins/1.x/proxies: Have a transparent upgrade proxy contract and a business logic contract. The business logic contract executes in the context of the upgrade proxy contract. The upgrade logic resides in the proxy contract. Issues with this approach are that extreme caution needs to be exercised to ensure there are no storage slot collisions between the proxy and the business logic contracts, and between different versions of the business logic contract.
+* Transparent Upgrade Proxy Pattern with Upgradable Upgrade Logic: As per the Transparent Upgrade Proxy Pattern described above, but with the upgrade logic in the business logic contract. The advantage of this approach is that governance that exists in the business logic contract can be used to approve the upgrade. Issues with this approach over and above the issues with the Transparent Upgrade Proxy Pattern approach is that bugs with the business logic contract can interfere with the upgrade logic, thus preventing upgrade, or enabling an attacker to maliciously upgrade the contract.
```

Comment:
Do we have a link we can add for this?

Reply:
The only instance I have seen of the last one is the Wormhole code.

Reply:
I could include a link to this, though am not sure this is helpful:
https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/Implementation.sol

Reply:
I see. Does it still make sense to list this as a "common methodology" if we only have a single instance we can cite?